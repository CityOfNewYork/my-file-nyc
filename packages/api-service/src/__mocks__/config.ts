const config = jest.requireActual('@/config')

export const EnvironmentVariable = config.EnvironmentVariable

export const getConfiguration = jest
  .fn()
  .mockImplementation((item) =>
    item != EnvironmentVariable.SENTRY_DSN ? item : undefined,
  )

export const isProduction = jest.fn().mockImplementation(() => false)

export const requireConfiguration = jest.fn().mockImplementation((item) => item)
