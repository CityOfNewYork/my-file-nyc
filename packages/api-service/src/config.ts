export enum EnvironmentVariable {
  NODE_ENV = 'NODE_ENV',
  DOCUMENTS_BUCKET = 'DOCUMENTS_BUCKET',
  USERINFO_ENDPOINT = 'USERINFO_ENDPOINT',
  WEB_APP_DOMAIN = 'WEB_APP_DOMAIN',
  EMAIL_SENDER = 'EMAIL_SENDER',
  AGENCY_EMAIL_DOMAINS_WHITELIST = 'AGENCY_EMAIL_DOMAINS_WHITELIST',
  CREATE_COLLECTION_ZIP_FUNCTION_NAME = 'CREATE_COLLECTION_ZIP_FUNCTION_NAME',
  ACTIVITY_RECORD_SQS_QUEUE_URL = 'ACTIVITY_RECORD_SQS_QUEUE_URL',
  ACTIVITY_CLOUDWATCH_LOG_GROUP = 'ACTIVITY_CLOUDWATCH_LOG_GROUP',
  EMAIL_PROCESSOR_SQS_QUEUE_URL = 'EMAIL_PROCESSOR_SQS_QUEUE_URL',
}

export const getConfiguration = (
  key: EnvironmentVariable,
): string | undefined => {
  return process.env[key]
}

export const isProduction = () =>
  getConfiguration(EnvironmentVariable.NODE_ENV) === 'production'

export const requireConfiguration = (key: EnvironmentVariable) => {
  const configuration = getConfiguration(key)
  if (configuration === undefined) {
    throw new Error(`Configuration key ${key} not found!`)
  }
  return configuration
}
