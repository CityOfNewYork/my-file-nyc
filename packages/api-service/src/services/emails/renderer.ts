import { render } from 'mustache'
import { existsSync, readFileSync, writeFileSync } from 'fs'
import { join } from 'path'

// set this to true to save the html files locally so they can be inspected for testing
const WRITE_FILES = false

export const templatePath = 'templates'

const loadedTemplates: { [index: string]: string } = {}

export const writeSnapshotFile = (
  name: string,
  content: string | undefined,
) => {
  if (WRITE_FILES) {
    writeFileSync(join(__dirname, '__snapshots__', name + '.html'), content)
  }
}

const loadTemplate = (templateName: string) => {
  if (!loadedTemplates[templateName]) {
    const templateResolvedPath = join(
      __dirname,
      templatePath,
      `${templateName}.mustache`,
    )
    if (!existsSync(templateResolvedPath)) {
      throw new Error(`${templateName} not found`)
    }
    loadedTemplates[templateName] = readFileSync(templateResolvedPath, 'utf8')
  }
  return loadedTemplates[templateName]
}

const partials = {
  header: loadTemplate('header'),
  footer: loadTemplate('footer'),
  horizontalRule: loadTemplate('horizontalRule'),
}

export const renderTemplate = (template: string, data: any) => {
  return render(loadTemplate(template), data, partials)
}
