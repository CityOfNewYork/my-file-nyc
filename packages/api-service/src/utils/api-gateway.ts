import {
  APIGatewayProxyEventV2,
  APIGatewayProxyStructuredResultV2,
} from 'aws-lambda'

export const getPathParameter = (
  event: APIGatewayProxyEventV2,
  parameterName: string,
): string | undefined => {
  return event.pathParameters && event.pathParameters[parameterName]
    ? event.pathParameters[parameterName]
    : undefined
}

export const getQueryStringParameter = (
  event: APIGatewayProxyEventV2,
  parameterName: string,
): string | undefined => {
  return event.queryStringParameters &&
    event.queryStringParameters[parameterName]
    ? event.queryStringParameters[parameterName]
    : undefined
}

export const getUrl = (event: APIGatewayProxyEventV2): string => {
  return event.requestContext.domainName
}

export const getUserId = (
  event: APIGatewayProxyEventV2,
): string | undefined => {
  if (!event.requestContext.authorizer || !event.requestContext.authorizer.jwt)
    return undefined
  return event.requestContext.authorizer.jwt.claims['sub'] as string
}

export const createJsonResponse = <T = any>(
  body: T,
  httpStatusCode = 200,
): APIGatewayProxyStructuredResultV2 => {
  return {
    cookies: [],
    isBase64Encoded: false,
    statusCode: httpStatusCode,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  }
}

export const createStatusCodeResponse = (
  httpStatusCode = 200,
): APIGatewayProxyStructuredResultV2 => {
  return {
    cookies: [],
    isBase64Encoded: false,
    statusCode: httpStatusCode,
  }
}

export const createErrorResponse = (msg: string, httpStatusCode = 400) =>
  createJsonResponse({ message: msg }, httpStatusCode)
