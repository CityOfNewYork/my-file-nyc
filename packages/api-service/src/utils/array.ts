export const hasValue = <TValue>(
  value: TValue | null | undefined,
): value is TValue => {
  return !!value
}
