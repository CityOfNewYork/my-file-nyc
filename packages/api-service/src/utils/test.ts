import { User } from '@/models/user'
import { requireUserData } from '@/services/users'
import { APIGatewayProxyEventV2, Context } from 'aws-lambda'
import { setContext } from './middleware'

export const toMockedFunction = <T extends (...args: any[]) => any>(
  input: T,
) => {
  return input as jest.MockedFunction<T>
}

export const importMock = async (modulePath: string) => {
  return (await import(modulePath)) as { [index: string]: jest.Mock }
}

export const setUserId = (userId: string, event: APIGatewayProxyEventV2) => {
  event.requestContext.authorizer = {
    jwt: {
      claims: {
        sub: userId,
        iat: 1605745134,
      },
      scopes: [],
    },
  }
  return event
}
export const createMockEvent = (
  data?: Partial<APIGatewayProxyEventV2>,
): APIGatewayProxyEventV2 => {
  return {
    headers: {
      authorization: 'my-token',
    },
    isBase64Encoded: false,
    rawPath: 'test',
    rawQueryString: 'test',
    requestContext: {
      accountId: '1111111111',
      apiId: 'apiId',
      domainName: 'domainName',
      domainPrefix: 'domainPrefix',
      http: {
        method: 'GET',
        path: 'path',
        protocol: 'HTTPS',
        sourceIp: '127.0.0.1',
        userAgent: 'Custom-UA',
      },
      requestId: '4A047E06-98D3-43AA-8501-424D8E0DEA32',
      routeKey: '4A047E06-98D3-43AA-8501-424D8E0DEA32',
      stage: 'stage',
      time: '2020-01-01T00:00:00.000Z',
      timeEpoch: 123456789,
    },
    routeKey: 'routeKey',
    version: '1',
    ...data,
  }
}

export const createMockContext = (): Context =>
  (jest.fn() as unknown) as Context

export const getObjectKeys = (obj: any): string[] => {
  return gatherKeysFromObject(obj)
}

const gatherKeysFromObject = (
  obj: any,
  keys: string[] = [],
  prefix = '',
): string[] => {
  if (obj === null || obj === undefined) {
    return keys
  } else if (Array.isArray(obj)) {
    obj.map((item, index) =>
      gatherKeysFromObject(item, keys, prefix + '[' + index + ']'),
    )
  } else if (obj.constructor === Object) {
    Object.entries(obj).map(([key, value]) =>
      gatherKeysFromObject(value, keys, prefix ? `${prefix}.${key}` : key),
    )
  } else if (prefix) {
    keys.push(prefix)
  }
  return keys
}

export const mockUserData = (
  userId: string,
  email = 'jcitizen@example.com',
  attributes?: any,
) => {
  const data = {
    id: userId,
    givenName: 'Jane',
    familyName: 'Citizen',
    email,
    attributes,
  }
  toMockedFunction(requireUserData).mockImplementation(
    async (request, requireTermsOfUseAcceptance) => {
      await setContext('userId', () => userId)(request)
      return User.fromJson(data)
    },
  )
  return data
}
