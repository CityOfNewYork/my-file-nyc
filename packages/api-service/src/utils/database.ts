import { Model } from 'objection'
import Knex from 'knex'
import knexConfig from '@/knexfile'
import { EnvironmentVariable, getConfiguration } from '@/config'

export const connectDatabase = (config?: string) => {
  config = getConfiguration(EnvironmentVariable.NODE_ENV) || 'development'
  const knex = Knex(knexConfig[config])
  Model.knex(knex)
  return knex
}
