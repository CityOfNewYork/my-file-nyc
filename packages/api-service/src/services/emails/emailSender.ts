/* eslint-disable @typescript-eslint/no-var-requires */
import SES from 'aws-sdk/clients/ses'
import {
  EnvironmentVariable,
  isProduction,
  requireConfiguration,
} from '@/config'
import { captureAWSClient } from 'aws-xray-sdk'
import { renderTemplate } from './renderer'

const ses = captureAWSClient(new SES())

type SendEmailOptions = {
  template: string
  subject: string
  data: any
  destination: SES.Destination
  forceSend?: boolean
}

export const sendEmail = async (opts: SendEmailOptions) => {
  const { template, subject, data, destination, forceSend = false } = opts
  const emailSender = requireConfiguration(EnvironmentVariable.EMAIL_SENDER)
  const webAppLogoSrc =
    'https://' +
    requireConfiguration(EnvironmentVariable.WEB_APP_DOMAIN) +
    '/images/city-logo.png'
  const body = renderTemplate(template, {
    ...data,
    webAppLogoSrc,
  })
  return new Promise<SES.SendEmailResponse & { Request: SES.SendEmailRequest }>(
    (resolve, reject) => {
      const request = {
        Destination: destination,
        Message: {
          Subject: {
            Data: subject,
            Charset: 'utf8',
          },
          Body: {
            Html: {
              Data: body,
              Charset: 'utf8',
            },
          },
        },
        Source: emailSender,
      }
      if (!forceSend && !isProduction()) {
        resolve({
          Request: request,
          MessageId: 'MOCKMESSAGEID',
        })
        return
      }
      ses.sendEmail(request, (err, data) => {
        if (err) reject(err)
        else
          resolve({
            Request: request,
            ...data,
          })
      })
    },
  )
}
