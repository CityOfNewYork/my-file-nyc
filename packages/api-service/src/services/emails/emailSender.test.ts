import { requireConfiguration } from '@/config'
import { toMockedFunction } from '@/utils/test'
import { sendEmail } from './emailSender'
import { writeSnapshotFile } from './renderer'

jest.mock('@/config')

describe('sendEmail', () => {
  beforeEach(() => {
    toMockedFunction(requireConfiguration).mockImplementationOnce(
      () => 'email-sender',
    )
    // set this to an actual domain to see images load
    toMockedFunction(requireConfiguration).mockImplementationOnce(
      () => 'mydatalocker.org',
    )
  })
  it('renders collectionSharedNotification template correctly', async () => {
    const results = await sendEmail({
      template: 'collectionSharedNotification',
      destination: {
        ToAddresses: ['testemail@example.com'],
      },
      subject: `You've been sent documents`,
      data: {
        link: 'https://mylink1234',
        name: 'Sam Smithers',
      },
    })
    expect(results.Request.Destination).toMatchInlineSnapshot(`
      Object {
        "ToAddresses": Array [
          "testemail@example.com",
        ],
      }
    `)
    expect(results.Request.Message.Body.Html?.Data).toMatchSnapshot()
    writeSnapshotFile(
      'collectionSharedNotification',
      results.Request.Message.Body.Html?.Data,
    )
  })
  it('renders delegateUserInvite template correctly', async () => {
    const results = await sendEmail({
      template: 'delegateUserInvite',
      destination: {
        ToAddresses: ['testemail@example.com'],
      },
      subject: `You've been granted access to an account`,
      data: {
        link: 'https://mylink1234',
        name: 'Sam Smithers',
      },
    })
    expect(results.Request.Destination).toMatchInlineSnapshot(`
      Object {
        "ToAddresses": Array [
          "testemail@example.com",
        ],
      }
    `)
    expect(results.Request.Message.Body.Html?.Data).toMatchSnapshot()
    writeSnapshotFile(
      'delegateUserInvite',
      results.Request.Message.Body.Html?.Data,
    )
  })
})
