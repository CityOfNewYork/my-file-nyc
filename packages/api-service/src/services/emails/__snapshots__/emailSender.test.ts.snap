// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendEmail renders collectionSharedNotification template correctly 2`] = `
"
<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Strict//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\\">
<html data-editor-version=\\"2\\" class=\\"sg-campaigns\\" xmlns=\\"http://www.w3.org/1999/xhtml\\">

<head>
  <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=utf-8\\" />
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, minimum-scale=1, maximum-scale=1\\" />
  <!--[if !mso]><!-->
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=Edge\\" />
  <!--<![endif]-->
  <!--[if (gte mso 9)|(IE)]>
    <xml>
    <o:OfficeDocumentSettings>
    <o:AllowPNG/>
    <o:PixelsPerInch>96</o:PixelsPerInch>
    </o:OfficeDocumentSettings>
    </xml>
    <![endif]-->
  <!--[if (gte mso 9)|(IE)]>
    <style type=\\"text/css\\">
      body {width: 515px;margin: 0 auto;}
      table {border-collapse: collapse;}
      table, td {mso-table-lspace: 0pt;mso-table-rspace: 0pt;}
      img {-ms-interpolation-mode: bicubic;}
    </style>
    <![endif]-->
    <style>
    @import url('https://fonts.googleapis.com/css2?family=Noto+Sans:wght@400;700&display=swap');
    </style>

  <style type=\\"text/css\\">
    body,
    p,
    div {
      font-family: 'Noto Sans', sans-serif;
      font-size: 14px;
    }

    h2 {
      font-style: normal;
      font-weight: 600;
      line-height: 27px;
      text-align: center;
      color: #2F3033;
    }

    body {
      color: #000000;
    }

    body a {
      text-decoration: none;
    }

    p {
      margin: 0;
      padding: 0;
    }

    table.wrapper {
      width: 100% !important;
      table-layout: fixed;
      -webkit-font-smoothing: antialiased;
      -webkit-text-size-adjust: 100%;
      -moz-text-size-adjust: 100%;
      -ms-text-size-adjust: 100%;
    }

    img.max-width {
      max-width: 100% !important;
    }

    .column.of-2 {
      width: 50%;
    }

    .column.of-3 {
      width: 33.333%;
    }

    .column.of-4 {
      width: 25%;
    }

    .primary-color {
      color: #2157E4
    }

    a.cta, div.cta > a {
        background-color: #2157E4;
        border: 1px solid #2157E4;
        border-radius: 8px;
        display: inline-block;
        color: #fff;
        padding: 9px 50px;
        text-align: center;
        text-decoration: none;
        -webkit-text-size-adjust: none;
    }

    [role=\\"modules-container\\"] {
      border-radius: 4px;
    }

    @media screen and (max-width:515px) {

      [role=\\"modules-container\\"] {
        border-radius: 0;
      }

      .preheader .rightColumnContent,
      .footer .rightColumnContent {
        text-align: left !important;
      }

      .preheader .rightColumnContent div,
      .preheader .rightColumnContent span,
      .footer .rightColumnContent div,
      .footer .rightColumnContent span {
        text-align: left !important;
      }

      .preheader .rightColumnContent,
      .preheader .leftColumnContent {
        font-size: 80% !important;
        padding: 5px 0;
      }

      table.wrapper-mobile {
        width: 100% !important;
        table-layout: fixed;
      }

      img.max-width {
        height: auto !important;
        max-width: 480px !important;
      }

      a.bulletproof-button {
        display: block !important;
        width: auto !important;
        font-size: 80%;
        padding-left: 0 !important;
        padding-right: 0 !important;
      }

      .columns {
        width: 100% !important;
      }

      .column {
        display: block !important;
        width: 100% !important;
        padding-left: 0 !important;
        padding-right: 0 !important;
        margin-left: 0 !important;
        margin-right: 0 !important;
      }

      .total_spacer {
        padding: 0px 0px 0px 0px;
      }
    }
  </style>
  <!--user entered Head Start-->

  <!--End Head user entered-->
</head>

<body>
  <center class=\\"wrapper\\"
    data-body-style=\\"font-size: 14px; font-family: arial; color: #000000; background-color: #ebebeb;\\">
    <div class=\\"webkit\\">
      <table cellpadding=\\"0\\" cellspacing=\\"0\\" border=\\"0\\" width=\\"100%\\" class=\\"wrapper\\" bgcolor=\\"#ebebeb\\">
        <tr>
          <td valign=\\"top\\" bgcolor=\\"#ebebeb\\" width=\\"100%\\">
            <table width=\\"100%\\" role=\\"content-container\\" class=\\"outer\\" align=\\"center\\" cellpadding=\\"0\\" cellspacing=\\"0\\"
              border=\\"0\\">
              <tr>
                <td width=\\"100%\\">
                  <table width=\\"100%\\" cellpadding=\\"0\\" cellspacing=\\"0\\" border=\\"0\\">
                    <tr>
                      <td>
                        <!--[if mso]>
                          <center>
                          <table><tr><td width=\\"600\\">
                          <![endif]-->
                        <table width=\\"100%\\" cellpadding=\\"0\\" cellspacing=\\"0\\" border=\\"0\\"
                          style=\\"width: 100%; max-width:515px; padding: 20px 0 20px 0;\\" align=\\"center\\">
                          <tr>
                            <td role=\\"modules-container\\"
                              style=\\"padding: 0px 0px 0px 0px; color: #000000; text-align: left;\\" bgcolor=\\"#ffffff\\"
                              width=\\"100%\\" align=\\"left\\">

                              <table class=\\"module preheader preheader-hide\\" role=\\"module\\" data-type=\\"preheader\\"
                                border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" width=\\"100%\\"
                                style=\\"display: none !important; mso-hide: all; visibility: hidden; opacity: 0; color: transparent; height: 0; width: 0;\\">
                                <tr>
                                  <td role=\\"module-content\\">
                                    <p></p>
                                  </td>
                                </tr>
                              </table>

                              <table class=\\"wrapper\\" role=\\"module\\" data-type=\\"image\\" border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" width=\\"100%\\" style=\\"table-layout: fixed;\\">
                                <tr>
                                  <td style=\\"font-size:6px;line-height:10px;padding:75px 0px 0px 0px;\\" valign=\\"top\\" align=\\"center\\">
                                    <img class=\\"max-width\\" border=\\"0\\" style=\\"display:block;color:#000000;text-decoration:none;font-family:Helvetica, arial, sans-serif;font-size:16px;max-width:115px !important;width:100%;height:auto !important;\\" src=\\"https://mydatalocker.org/images/city-logo.png\\" alt=\\"\\" width=\\"115\\">
                                  </td>
                                </tr>
                              </table>

                              <table class=\\"module\\" role=\\"module\\" data-type=\\"text\\" border=\\"0\\" cellpadding=\\"0\\"
                                cellspacing=\\"0\\" width=\\"100%\\" style=\\"table-layout: fixed;\\">
                                <tr>
                                  <td style=\\"padding:25px 20px 25px 20px;line-height:20px;text-align:inherit;\\"
                                    height=\\"100%\\" valign=\\"top\\" bgcolor=\\"\\">

      <h2 style=\\"text-align: center;\\">You've received new <br />documents.</h2>
      <br/>
      <table width=\\"100%\\" border=\\"0\\" cellspacing=\\"0\\" cellpadding=\\"0\\">
          <tr>
              <td align=\\"center\\">
                  <div style=\\"max-width: 340px; border-top: 1px solid #D8DBDF\\"></div>
              </td>
          </tr>
      </table>
      <br/>
      <p style=\\"text-align: center; font-weight: bold; padding-top: 20px\\" class=\\"primary-color\\">Sam Smithers</p>
      <br/>
      <p style=\\"text-align: center;\\">has shared documents with you in the <br/>Datalocker. Click below to view them. </p>
      <br/>
      <table width=\\"100%\\" border=\\"0\\" cellspacing=\\"0\\" cellpadding=\\"0\\">
          <tr>
              <td align=\\"center\\">
                  <div>
                      <a href=\\"https:&#x2F;&#x2F;mylink1234\\" class=\\"cta\\">View Files</a>
                  </div>
              </td>
          </tr>
      </table>


                                    <br/><br/>
                                  </td>
                                </tr>
                              </table>
                              <table class=\\"module\\" role=\\"module\\" data-type=\\"divider\\" border=\\"0\\" cellpadding=\\"0\\"
                                cellspacing=\\"0\\" width=\\"100%\\" style=\\"table-layout: fixed;\\">
                                <tr>
                                  <td style=\\"padding:0px 0px 0px 0px;\\" role=\\"module-content\\" height=\\"100%\\" valign=\\"top\\"
                                    bgcolor=\\"\\">
                                    <table border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" align=\\"center\\" width=\\"100%\\"
                                      height=\\"5px\\" style=\\"line-height:5px; font-size:5px;\\">
                                      <tr>
                                        <td style=\\"padding: 0px 0px 5px 0px;\\" bgcolor=\\"#ebebeb\\"></td>
                                      </tr>
                                    </table>
                                  </td>
                                </tr>
                              </table>

                            </td>
                          </tr>
                        </table>
                        <!--[if mso]>
                          </td></tr></table>
                          </center>
                          <![endif]-->
                      </td>
                    </tr>
                  </table>
                </td>
              </tr>
            </table>
          </td>
        </tr>
      </table>
    </div>
  </center>
</body>

</html>

"
`;

exports[`sendEmail renders delegateUserInvite template correctly 2`] = `
"
<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Strict//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\\">
<html data-editor-version=\\"2\\" class=\\"sg-campaigns\\" xmlns=\\"http://www.w3.org/1999/xhtml\\">

<head>
  <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=utf-8\\" />
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, minimum-scale=1, maximum-scale=1\\" />
  <!--[if !mso]><!-->
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=Edge\\" />
  <!--<![endif]-->
  <!--[if (gte mso 9)|(IE)]>
    <xml>
    <o:OfficeDocumentSettings>
    <o:AllowPNG/>
    <o:PixelsPerInch>96</o:PixelsPerInch>
    </o:OfficeDocumentSettings>
    </xml>
    <![endif]-->
  <!--[if (gte mso 9)|(IE)]>
    <style type=\\"text/css\\">
      body {width: 515px;margin: 0 auto;}
      table {border-collapse: collapse;}
      table, td {mso-table-lspace: 0pt;mso-table-rspace: 0pt;}
      img {-ms-interpolation-mode: bicubic;}
    </style>
    <![endif]-->
    <style>
    @import url('https://fonts.googleapis.com/css2?family=Noto+Sans:wght@400;700&display=swap');
    </style>

  <style type=\\"text/css\\">
    body,
    p,
    div {
      font-family: 'Noto Sans', sans-serif;
      font-size: 14px;
    }

    h2 {
      font-style: normal;
      font-weight: 600;
      line-height: 27px;
      text-align: center;
      color: #2F3033;
    }

    body {
      color: #000000;
    }

    body a {
      text-decoration: none;
    }

    p {
      margin: 0;
      padding: 0;
    }

    table.wrapper {
      width: 100% !important;
      table-layout: fixed;
      -webkit-font-smoothing: antialiased;
      -webkit-text-size-adjust: 100%;
      -moz-text-size-adjust: 100%;
      -ms-text-size-adjust: 100%;
    }

    img.max-width {
      max-width: 100% !important;
    }

    .column.of-2 {
      width: 50%;
    }

    .column.of-3 {
      width: 33.333%;
    }

    .column.of-4 {
      width: 25%;
    }

    .primary-color {
      color: #2157E4
    }

    a.cta, div.cta > a {
        background-color: #2157E4;
        border: 1px solid #2157E4;
        border-radius: 8px;
        display: inline-block;
        color: #fff;
        padding: 9px 50px;
        text-align: center;
        text-decoration: none;
        -webkit-text-size-adjust: none;
    }

    [role=\\"modules-container\\"] {
      border-radius: 4px;
    }

    @media screen and (max-width:515px) {

      [role=\\"modules-container\\"] {
        border-radius: 0;
      }

      .preheader .rightColumnContent,
      .footer .rightColumnContent {
        text-align: left !important;
      }

      .preheader .rightColumnContent div,
      .preheader .rightColumnContent span,
      .footer .rightColumnContent div,
      .footer .rightColumnContent span {
        text-align: left !important;
      }

      .preheader .rightColumnContent,
      .preheader .leftColumnContent {
        font-size: 80% !important;
        padding: 5px 0;
      }

      table.wrapper-mobile {
        width: 100% !important;
        table-layout: fixed;
      }

      img.max-width {
        height: auto !important;
        max-width: 480px !important;
      }

      a.bulletproof-button {
        display: block !important;
        width: auto !important;
        font-size: 80%;
        padding-left: 0 !important;
        padding-right: 0 !important;
      }

      .columns {
        width: 100% !important;
      }

      .column {
        display: block !important;
        width: 100% !important;
        padding-left: 0 !important;
        padding-right: 0 !important;
        margin-left: 0 !important;
        margin-right: 0 !important;
      }

      .total_spacer {
        padding: 0px 0px 0px 0px;
      }
    }
  </style>
  <!--user entered Head Start-->

  <!--End Head user entered-->
</head>

<body>
  <center class=\\"wrapper\\"
    data-body-style=\\"font-size: 14px; font-family: arial; color: #000000; background-color: #ebebeb;\\">
    <div class=\\"webkit\\">
      <table cellpadding=\\"0\\" cellspacing=\\"0\\" border=\\"0\\" width=\\"100%\\" class=\\"wrapper\\" bgcolor=\\"#ebebeb\\">
        <tr>
          <td valign=\\"top\\" bgcolor=\\"#ebebeb\\" width=\\"100%\\">
            <table width=\\"100%\\" role=\\"content-container\\" class=\\"outer\\" align=\\"center\\" cellpadding=\\"0\\" cellspacing=\\"0\\"
              border=\\"0\\">
              <tr>
                <td width=\\"100%\\">
                  <table width=\\"100%\\" cellpadding=\\"0\\" cellspacing=\\"0\\" border=\\"0\\">
                    <tr>
                      <td>
                        <!--[if mso]>
                          <center>
                          <table><tr><td width=\\"600\\">
                          <![endif]-->
                        <table width=\\"100%\\" cellpadding=\\"0\\" cellspacing=\\"0\\" border=\\"0\\"
                          style=\\"width: 100%; max-width:515px; padding: 20px 0 20px 0;\\" align=\\"center\\">
                          <tr>
                            <td role=\\"modules-container\\"
                              style=\\"padding: 0px 0px 0px 0px; color: #000000; text-align: left;\\" bgcolor=\\"#ffffff\\"
                              width=\\"100%\\" align=\\"left\\">

                              <table class=\\"module preheader preheader-hide\\" role=\\"module\\" data-type=\\"preheader\\"
                                border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" width=\\"100%\\"
                                style=\\"display: none !important; mso-hide: all; visibility: hidden; opacity: 0; color: transparent; height: 0; width: 0;\\">
                                <tr>
                                  <td role=\\"module-content\\">
                                    <p></p>
                                  </td>
                                </tr>
                              </table>

                              <table class=\\"wrapper\\" role=\\"module\\" data-type=\\"image\\" border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" width=\\"100%\\" style=\\"table-layout: fixed;\\">
                                <tr>
                                  <td style=\\"font-size:6px;line-height:10px;padding:75px 0px 0px 0px;\\" valign=\\"top\\" align=\\"center\\">
                                    <img class=\\"max-width\\" border=\\"0\\" style=\\"display:block;color:#000000;text-decoration:none;font-family:Helvetica, arial, sans-serif;font-size:16px;max-width:115px !important;width:100%;height:auto !important;\\" src=\\"https://mydatalocker.org/images/city-logo.png\\" alt=\\"\\" width=\\"115\\">
                                  </td>
                                </tr>
                              </table>

                              <table class=\\"module\\" role=\\"module\\" data-type=\\"text\\" border=\\"0\\" cellpadding=\\"0\\"
                                cellspacing=\\"0\\" width=\\"100%\\" style=\\"table-layout: fixed;\\">
                                <tr>
                                  <td style=\\"padding:25px 20px 25px 20px;line-height:20px;text-align:inherit;\\"
                                    height=\\"100%\\" valign=\\"top\\" bgcolor=\\"\\">

      <h2 style=\\"text-align: center;\\">A client has granted access <br />to their account.</h2>
      <br/>
      <table width=\\"100%\\" border=\\"0\\" cellspacing=\\"0\\" cellpadding=\\"0\\">
          <tr>
              <td align=\\"center\\">
                  <div style=\\"max-width: 340px; border-top: 1px solid #D8DBDF\\"></div>
              </td>
          </tr>
      </table>
      <br/>
      <p style=\\"text-align: center; font-weight: bold; padding-top: 20px\\" class=\\"primary-color\\">Sam Smithers</p>
      <br/>
      <p style=\\"text-align: center;\\">has allowed you to manage and share files on<br/>their behalf in Datalocker.</p>
      <br/>
      <table width=\\"100%\\" border=\\"0\\" cellspacing=\\"0\\" cellpadding=\\"0\\">
          <tr>
              <td align=\\"center\\">
                  <div>
                      <a href=\\"https:&#x2F;&#x2F;mylink1234\\" class=\\"cta\\">Accept</a>
                  </div>
              </td>
          </tr>
      </table>


                                    <br/><br/>
                                  </td>
                                </tr>
                              </table>
                              <table class=\\"module\\" role=\\"module\\" data-type=\\"divider\\" border=\\"0\\" cellpadding=\\"0\\"
                                cellspacing=\\"0\\" width=\\"100%\\" style=\\"table-layout: fixed;\\">
                                <tr>
                                  <td style=\\"padding:0px 0px 0px 0px;\\" role=\\"module-content\\" height=\\"100%\\" valign=\\"top\\"
                                    bgcolor=\\"\\">
                                    <table border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" align=\\"center\\" width=\\"100%\\"
                                      height=\\"5px\\" style=\\"line-height:5px; font-size:5px;\\">
                                      <tr>
                                        <td style=\\"padding: 0px 0px 5px 0px;\\" bgcolor=\\"#ebebeb\\"></td>
                                      </tr>
                                    </table>
                                  </td>
                                </tr>
                              </table>

                            </td>
                          </tr>
                        </table>
                        <!--[if mso]>
                          </td></tr></table>
                          </center>
                          <![endif]-->
                      </td>
                    </tr>
                  </table>
                </td>
              </tr>
            </table>
          </td>
        </tr>
      </table>
    </div>
  </center>
</body>

</html>

"
`;
